main/exercise/ptest_category.class.php

```diff
@@ -0,0 +1,1174 @@
+                    $newCategoryList[$categoryId] = $categories[$categoryId];
+                }
+            }
+
+            $checkQuestionsWithNoCategory = false;
+            foreach ($categoriesAddedInExercise as $category) {
+                if (empty($category['category_id'])) {
+                    // Check
+                    $checkQuestionsWithNoCategory = true;
+                    break;
+                }
+            }
+
+            // Select questions that don't have any category related
+            if ($checkQuestionsWithNoCategory) {
+                $originalQuestionList = $check_in_question_list;
+                foreach ($originalQuestionList as $questionId) {
+                    $categoriesFlatten = array_flatten($categories);
+                    if (!in_array($questionId, $categoriesFlatten)) {
+                        $newCategoryList[0][] = $questionId;
+                    }
+                }
+            }
+            $categories = $newCategoryList;
+        }
+
+        return $categories;
+    }
+
+    /**
+     * Returns an array of $numberElements from $array.
+     *
+     * @param array
+     * @param int
+     *
+     * @return array
+     */
+    public static function getNElementsFromArray($array, $numberElements)
+    {
+        $list = $array;
+        shuffle($list);
+        if ($numberElements < count($list)) {
+            $list = array_slice($list, 0, $numberElements);
+        }
+
+        return $list;
+    }
+
+    /**
+     * @param int $questionId
+     * @param int $displayCategoryName
+     */
+    public static function displayCategoryAndTitle($questionId, $displayCategoryName = 1)
+    {
+        echo self::returnCategoryAndTitle($questionId, $displayCategoryName);
+    }
+
+    /**
+     * @param int $questionId
+     * @param int $in_display_category_name
+     *
+     * @return string|null
+     */
+    public static function returnCategoryAndTitle($questionId, $in_display_category_name = 1)
+    {
+        $is_student = !(api_is_allowed_to_edit(null, true) || api_is_session_admin());
+        $objExercise = Session::read('objExercise');
+        if (!empty($objExercise)) {
+            $in_display_category_name = $objExercise->display_category_name;
+        }
+        $content = null;
+        if (self::getCategoryNameForQuestion($questionId) != '' &&
+            ($in_display_category_name == 1 || !$is_student)
+        ) {
+            $content .= '<div class="page-header">';
+            $content .= '<h4>'.get_lang('Category').": ".self::getCategoryNameForQuestion($questionId).'</h4>';
+            $content .= "</div>";
+        }
+
+        return $content;
+    }
+
+    /**
+     * sortTabByBracketLabel ($tabCategoryQuestions)
+     * key of $tabCategoryQuestions are the category id (0 for not in a category)
+     * value is the array of question id of this category
+     * Sort question by Category.
+     */
+    public static function sortTabByBracketLabel($in_tab)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The variable $in_tab is not named in camelCase.](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5018478)